File: javaapp/src/main/java/math/sumofmultiples/SumOfMultiples.java
package java.math.sumofmultiples;

import java.util.stream.IntStream;

public class SumOfMultiples {

    public int Sum(int number, int[] multiples) {
        //create range for the numbers
        if (multiples.length == 0) {
            return 0;
        }
        IntStream numbers = IntStream.rangeClosed(0, number - 1);
        if (multiples.length == 3) {
            return numbers
                    .filter(n -> n % multiples[0] == 0 || n % multiples[1] == 0 || n % multiples[2] == 0)
                    .sum();
        }

        return numbers.filter(n -> n % multiples[0] == 0 || n % multiples[1] == 0).sum();
    }
}

File: app/src/main/kotlin/com/kotlinapp/ktmath/sumOfMultiples/SumOfMultiples.kt
object SumOfMultiples{

    fun sum(numbers: Set<Int> = emptySet(), limit: Int) : Int{
        if( limit <= 1){
            return 0
        }

        val range = IntRange(1, limit -1)
        val multiples = arrayListOf<Int>()

        for (num in numbers){
            range
                    .asSequence()
                    .filter { it % num == 0 && it !in multiples }
                    .forEach { multiples.add(it) }
        }

        return multiples.sum()
    }


    fun sumV2(numbers: Set<Int> = emptySet(), limit: Int) =
            (1 until limit)
                    .filter {
                        value -> numbers.any{ it.hasMultiple(value) }
                    }.sum()


    private fun Int.hasMultiple(number : Int) = (number >= this) && (number % this == 0)
}